[Validator] check for empty host when calling checkdnsrr()

apetitpa opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master|
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #22106 |
| License | MIT |
| Doc PR | n/a |

This should fix the email validator issue

Commit:
[Validator] Allow checkMX() to return false when $host is empty

| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #22106 |
| License | MIT |
| Doc PR | n/a |

apetitpa commented:
Sorry for the last commit, I rebased branch with master and didn't notice you already did.

apetitpa commented:
The hhvm build failed but it does not seem to be related to this PR.

fabpot commented:
The regex should not be changed. But the condition should indeed be added in other methods using checkdnsrr.

apetitpa commented:
thank you @fabpot and @jordscream I'll make some modifications, add unit tests and rebase my PR wwith 2.7

xabbuh commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -118,6 +118,10 @@ public function validate($value, Constraint $constraint)
      */
     private function checkMX($host)
     {
+        if (null === $host || '' === $host) {
```
`$host` can never be `null`, can it?

apetitpa replied:
I think you're right, I'll fix this. Thank you

xabbuh reviewed (changes requested):
Can you please also add a test to prevent future regressions?

Commit:
Add empty check on host in other methods + add unit tests

Add empty on host in other methods where checkdnsrr is called and add unit tests to prevent future regressions

Commit:
Remove malformed EmailValidatorTest + Update UrlValidator test

Remove malformed EmailValidatorTest + Update UrlValidator test

Commit:
fix coding standard to comply with fabbot

Commit:
Use LF line separator

jordscream commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -118,6 +118,10 @@ public function validate($value, Constraint $constraint)
+        if ('' === $host) {
+            return false;
+        }
+
```
preferred empty host condition in return statement

jordscream commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -130,6 +134,10 @@ private function checkMX($host)
+            return false;
+        }
+
         return $this->checkMX($host) || (checkdnsrr($host, 'A') || checkdnsrr($host, 'AAAA'));
```
preferred empty host condition in return statement

jsamouh reviewed (commented):
What about unit test for the main issue ? ^^

Commit:
Move empty condition in return statement

apetitpa commented:
About the unit test for the main issue, since checkMX() and checkHost() are both private how can I test it ?

jordscream commented:
@apetitpa just test a use case  as "test@test@" or "test@" in the EmailValidator. It should return the appropriate violation

Commit:
Add a test case to prevent future regressions

apetitpa commented:
Thank you @jordscream I forgot that the original issue was with emails containing @ twice

Nek- commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -118,7 +118,7 @@ public function validate($value, Constraint $constraint)
     private function checkMX($host)
     {
-        return checkdnsrr($host, 'MX');
+        return ('' !== $host) && checkdnsrr($host, 'MX');
```
why not `empty` check?

apetitpa replied:
Well there is a similar check line 45 `if (null === $value || '' === $value)` that does not use the native empty function so I assumed I should do the same but maybe I was wrong ?

fabpot replied:
We try to avoid using `empty` in Symfony.

fabpot replied:
parentheses can be removed

apetitpa replied:
I'll fix this in a minute thank you.

Nek- replied:
@fabpot interesting, why? (is the reason documented somewhere?)

fabpot replied:
`empty` is just too broad. For instance, it includes the `0` string as being `false`. Sometimes, that not desirable. So, we prefer to be very explicit about our expectations.

Commit:
Remove unnecessary parentheses

jordscream commented:
@apetitpa , it seems your unit test fails cause you do not check the appropriate violation

apetitpa commented:
@jordscream I think you're right I should check for MX_CHECK_FAILED_ERROR or HOST_CHECK_FAILED_ERROR

ro0NL commented on src/Symfony/Component/Validator/Constraints/UrlValidator.php:
```diff
@@ -72,7 +72,7 @@ public function validate($value, Constraint $constraint)
             $host = parse_url($value, PHP_URL_HOST);
 
-            if (!checkdnsrr($host, 'ANY')) {
+            if ('' === $host || !checkdnsrr($host, 'ANY')) {
```
Perhaps switch to `!is_string($host)` due possible `false` or `null` return values.

apetitpa replied:
I think this is a good idea. I'll do it thank you.

jordscream commented:
@apetitpa . When you think to do it ? ^^. Need this fix for a customer :-P

Commit:
Switch to `is_string` native method

Commit:
remove non relevant test case

apetitpa commented:
@jordscream I hope this week-end but to be honest I'm facing difficulties creating an effective and working test for this issue. Some help would be appreciated.

xabbuh commented:
> Dont think it needs to be dns-senstive though, as there will be no network traffic for `foo@bar@`.

Agreed, the issue to fix is that the DNS functions should in fact not be called anymore.

Commit:
Switch to `empty` native function to check emptiness

Switch to `empty` because depending on the PHP version, substr returns '' or false

Commit:
Add a test to prevent future regressions

ro0NL commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -118,7 +118,7 @@ public function validate($value, Constraint $constraint)
     private function checkMX($host)
     {
-        return checkdnsrr($host, 'MX');
+        return !empty($host) && checkdnsrr($host, 'MX');
```
Please avoid `empty` :) We expect a string, and only a empty string (`$host === ''`)  should not be passed to `checkdnsrr`.

apetitpa replied:
I know but `substr($value, strrpos($value, '@') + 1)` with `$value = foo@bar.fr@` returns `""` with PHP 7.* and `false` with PHP 5.6. So only checking for string emptiness is not enough.

ro0NL replied:
Let's cast the substring result above (or perhaps explicitly check for false); as currently we literally get `string('false')` as a replacement for `{{ value }}` in the violations above; due https://github.com/apetitpa/symfony/blob/master/src/Symfony/Component/Validator/ConstraintValidator.php#L139

So we got an extra (tiny) bug here ;-)

apetitpa replied:
It's done, thank you

ro0NL commented on src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:
```diff
@@ -229,4 +229,32 @@ public function testHostnameIsProperlyParsed()
         $this->assertNoViolation();
     }
+
+    public function getCheckTypes()
```
Should be `provideCheckTypes` i guess, but it's a tiny semantic ;)

apetitpa replied:
I'll make the modification thank you

Commit:
rename dataset provider

Commit:
cast substr result to string and remove empty function use

Commit:
update dataProvider function name

apetitpa commented:
@xabbuh is it ok now ?

fabpot commented on src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:
```diff
@@ -229,4 +229,32 @@ public function testHostnameIsProperlyParsed()
         $this->assertNoViolation();
     }
+
+    public function provideCheckTypes()
```
Can you move the provider after the test (that's the convention we are using in Symfony).

apetitpa replied:
Of course, will do in a minute.

Commit:
move provider after test

xabbuh reviewed (approved):
👍

fabpot commented:
Thank you @apetitpa.
